modules/programs/sbt.nix

```diff
@@ -0,0 +1,122 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  renderPlugin = plugin: ''
+    addSbtPlugin("${plugin.org}" % "${plugin.artifact}" % "${plugin.version}")
+  '';
+
+  renderCredential = cred: ''
+    credentials += Credentials("${cred.realm}", "${cred.host}", "${cred.user}", "${cred.password}")
+  '';
+
+  sbtTypes = {
+    plugin = types.submodule {
+      options = {
+        org = mkOption {
+          type = types.str;
+          description = "The organization the artifact is published under";
+        };
+        artifact = mkOption {
+          type = types.str;
+          description = "The name of the artifact";
+        };
+        version = mkOption {
+          type = types.str;
+          description = "The version of the plugin";
+        };
+      };
+    };
+
+    credential = types.submodule {
+      options = {
+        realm = mkOption {
+          type = types.str;
+          description = "The realm of the repository you're authenticating to";
+        };
+        host = mkOption {
+          type = types.str;
+          description =
+            "The hostname of the repository you're authenticating to";
+        };
+        user = mkOption {
+          type = types.str;
+          description = "The user you're using to authenticate";
+        };
+        password = mkOption {
```

Comment:
It is a bad idea to have plain text passwords since they will end up in the world-readable Nix store. In the email infrastructure we instead use `passwordCommand`. Perhaps that would work here as well? Something like

```
import sys.process._
…
credentials += Credentials("${cred.realm}", "${cred.host}", "${cred.user}", "${cred.passwordCommand}".!!)
```

Reply:
That does seem possible... so if I understand correctly, a user of this could put the password-printing command in a nix-untracked file and pass the path to it here?

In that case, wouldn't the path also end up in the store, with its contents (including the secret)?

Reply:
Apparently if I use `toString` instead of interpolating the command's path directly, it doesn't get copied into the store:

```diff
diff --git a/scala/default.nix b/scala/default.nix
index 5ca9c8a..9ad3e54 100644
--- a/scala/default.nix
+++ b/scala/default.nix
@@ -28,5 +28,12 @@
         version = "0.4.0";
       };
     in [ dependencyGraph projectGraph ];
+    credentials = [{
+      realm = "Foo";
+      host = "example.com";
+      user = "user";
+      passwordCommand =
+        "${pkgs.bash}/bin/bash ${toString ./examplecom-passwd.sh}";
+    }];
   };
 }
diff --git a/scala/examplecom-passwd.sh b/scala/examplecom-passwd.sh
new file mode 100755
index 0000000..b8a9523
--- /dev/null
+++ b/scala/examplecom-passwd.sh
@@ -0,0 +1,2 @@
+#!/bin/env bash
+echo -n "test"
```

This would probably need to be carefully documented.

Reply:
BTW. Please let me know if you would change something about the way I pass the shell there. Using nix-shell would add too much overhead IMO.

Reply:
Typically you would use a command that fetches the password from your chosen password storage. E.g., the system keyring or something like [pass](https://www.passwordstore.org/). For example in my email setup I have

    passwordCommand = "pass show email/robert@rycee.net";

which would fetch the password from a GPG encrypted file, indeed not only is it at all times encrypted on disk, I also have to plug in my yubikey to make the GPG key available to the pass command.

Reply:
By the way, have a look at the definition of the `passwordCommand` option in `email.nix` and also the [generated documentation](https://rycee.gitlab.io/home-manager/options.html#opt-accounts.email.accounts._name_.passwordCommand).
https://github.com/nix-community/home-manager/blob/98d030f723e0a4a446e56b276573efb8bef422f5/modules/accounts/email.nix#L232-L241

Reply:
> I also have to plug in my yubikey to make the GPG key available to the pass command.

Sorry, hijacking to ask an unrelated question: how do you get gpg to work with your key when you plug it in? I always have to do `gpg --card-status` or something like that for my key to work the first time I use it.

Reply:
> Typically you would use a command that fetches the password from your chosen password storage. E.g., the system keyring or something like pass

My password manager of choice is 1password, and any CLIs I might have will be much slower than what I want in the sbt startup, so I'll probably stick to git-crypted files in a git repository 😅

Reply:
@berbiche I think that is a regression, I have been using it for a few years and recently (a month ago or so, I started seeing this annoying behavior). Not sure where it comes from and exactly when it started to occur.

Reply:
> @berbiche I think that is a regression, I have been using it for a few years and recently (a month ago or so, I started seeing this annoying behavior). Not sure where it comes from and exactly when it started to occur.

Changing to `pinentry-qt` in my gpg-agent conf somehow resolved the issue. It seems to be a bug with `pinentry-gnome3`.